nx/lib/nx/defn/evaluator.ex

```diff
@@ -23,64 +23,101 @@ defmodule Nx.Defn.Evaluator do
   end
 
   @impl true
-  def __jit__(_key, vars, fun, _opts) do
+  def __jit__(_key, vars, fun, opts) do
+    hooks = Keyword.get(opts, :hooks, %{})
+
     fun.(vars)
-    |> composite_eval(vars, %{})
+    |> composite_eval(%{vars: vars, hooks: hooks}, %{})
     |> elem(0)
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :fun, args: [args, expr, _mfa]}}, _vars, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: :fun, args: [args, expr, _mfa]}}, state, cache) do
     fun =
       case length(args) do
-        1 -> fn arg1 -> expr |> composite_eval([arg1], %{}) |> elem(0) end
-        2 -> fn arg1, arg2 -> expr |> composite_eval([arg1, arg2], %{}) |> elem(0) end
+        1 ->
+          fn arg1 ->
+            expr |> composite_eval(%{state | vars: [arg1]}, %{}) |> elem(0)
+          end
+
+        2 ->
+          fn arg1, arg2 ->
+            expr |> composite_eval(%{state | vars: [arg1, arg2]}, %{}) |> elem(0)
```

Comment:
```suggestion
            {result_expr, _cache} = composite_eval(expr, %{state | vars: [arg1]}, %{})
            result_expr
          end

        2 ->
          fn arg1, arg2 ->
            {result_expr, _cache} = composite_eval(expr, %{state | vars: [arg1, arg2]}, %{})
            result_expr
```

This is mostly cosmetic, but I'm leaving the suggestion because this is something that I had to stop and think about